### PR TITLE
Fix cubed sphere velocity halo exchange at/near corners

### PR DESCRIPTION
**Pull request by ali-ramadhan**
Hijacking @christophernhill's branch and work!

@jm-c set up the same velocity halo exchange from https://github.com/CliMA/Oceananigans.jl/blob/7cc147bbf831fe4de66584d812e56debcfa6e43f/test/test_cubed_sphere_halo_exchange.jl#L378-L382 in MITgcm from which we were able to figure out where the Oceananigans cubed sphere halo exchange was going wrong.

There were a several discrepancies (all at halo corners or adjacent to the halo corners).

```
u face 1: 2 discrepancies
v face 1: 3 discrepancies
u face 2: 2 discrepancies
v face 2: 3 discrepancies
u face 3: 1 discrepancies
v face 3: 2 discrepancies
u face 4: 1 discrepancies
v face 4: 3 discrepancies
u face 5: 0 discrepancies
v face 5: 2 discrepancies
u face 6: 0 discrepancies
v face 6: 2 discrepancies
u: 6 discrepancies
v: 15 discrepancies
```

Not totally sure why our halo exchange algorithm isn't filling these correctly so for now we fix each one after filling all the velocity halos.

Now that our halo exchange exactly matches MITgcm's we might want to add some extra tests to make sure we're testing each of these points. Some existing tests need to be cleaned up too.

After this fix the Rossby-Haurwitz validation seems pretty happy (it used to exhibit numerical artifacts at some corners then blow up after a few days).

https://user-images.githubusercontent.com/20099589/118747990-0317e680-b829-11eb-94c3-40002f928e52.mp4

The eddying aquaplanet validation from PR #1602 doesn't look as happy so it might still have setup issues (and maybe also plotting issues since vorticity is at the corners).

### CODE REVIEW

**Review by christophernhill — APPROVED** [review]
@ali-ramadhan this looks great. I haven't tried it, but will now. I think given that things don't work well now then it would be good to merge as soon as we can.

**Comment by christophernhill** [comment]
For future reference. 

This PR .

- adds checks of velocity values in overlaps at the 3 tile corners, such as the example for a cube  face 1,3,5 corner shown in the figure.
- adds a temporary fix (that will need revisiting for MPI based work) to ensure the corner overlap velocity points are set correctly.

image

File: src/CubedSpheres/cubed_sphere_halo_filling.jl [hunk 1]
```diff
@@ -143,5 +143,48 @@ function fill_horizontal_velocity_halos!(u::CubedSphereField, v::CubedSphereFiel
         end
     end
 
+    # Manually handle corners to exactly match MITgcm halo filling algorithm.
+    # Not sure whether every single one of these is needed though.
+
+    Nx, Ny, Nz, Nf = size(u.grid)
+
+    @inbounds begin
```
Comment (glwagner) [inline src/CubedSpheres/cubed_sphere_halo_filling.jl:151]: Some comments about the local geometry (boundary and rotation of data source relative to destination) could help when prototyping an adjacency function in the future.

**Comment by christophernhill** [comment]
@ali-ramadhan GPU()!!! Very exciting 🚀

**Pull request merged**